### ISSUE TITLE
Will a softmax with focal loss be implemented? 

### ISSUE THREAD

**Issue by GeorgeSeif** [tensorflow/models#4245]
Please go to Stack Overflow for help and support:

http://stackoverflow.com/questions/tagged/tensorflow

Also, please understand that many of the models included in this repository are experimental and research-style code. If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **What is the top-level directory of the model you are using**: models/research/object_detection
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Trying to, see below
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: used pip install
- **TensorFlow version (use command below)**: 1.8.0
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: 
- **GPU model and memory**: K80 with 12GB
- **Exact command to reproduce**: See below

Thanks for the awesome Object Detection API!

I wanted to be able to use the [focal loss](http://openaccess.thecvf.com/content_ICCV_2017/papers/Lin_Focal_Loss_for_ICCV_2017_paper.pdf) with a softmax classification loss so I can apply it to R-FCN and Faster-RCNN based models.

Will there ever be any implementation for this? I tried doing it myself within the `WeightedSoftmaxClassificationLoss` function in `losses.py` like below:
```python
def _compute_loss(self, prediction_tensor, target_tensor, weights):
    """Compute loss function.

    Args:
      prediction_tensor: A float tensor of shape [batch_size, num_anchors,
        num_classes] representing the predicted logits for each class
      target_tensor: A float tensor of shape [batch_size, num_anchors,
        num_classes] representing one-hot encoded classification targets
      weights: a float tensor of shape [batch_size, num_anchors]

    Returns:
      loss: a float tensor of shape [batch_size, num_anchors]
        representing the value of the loss function.
    """
    num_classes = prediction_tensor.get_shape().as_list()[-1]
    prediction_tensor = tf.divide(
        prediction_tensor, self._logit_scale, name='scale_logit')
    per_row_cross_ent = (tf.nn.softmax_cross_entropy_with_logits(
        labels=tf.reshape(target_tensor, [-1, num_classes]),
        logits=tf.reshape(prediction_tensor, [-1, num_classes])))

    ####################################################
    # FOCAL LOSS PART START
    ####################################################
    prediction_probabilities = tf.nn.softmax(prediction_tensor)
    p_t = ((target_tensor * prediction_probabilities) +
           ((1 - target_tensor) * (1 - prediction_probabilities)))

    modulating_factor = 1.0
    if self._gamma:
      modulating_factor = tf.pow(1.0 - p_t, self._gamma)
    alpha_weight_factor = 1.0
    if self._alpha is not None:
      alpha_weight_factor = (target_tensor * self._alpha +
                             (1 - target_tensor) * (1 - self._alpha))
    focal_cross_entropy_loss = (modulating_factor * alpha_weight_factor *
                                per_row_cross_ent)

    ####################################################
    # FOCAL LOSS PART END
    ####################################################
    return tf.reshape(focal_cross_entropy_loss, tf.shape(weights)) * weights
``` 

My idea was to follow the same style as the already implemented Sigmoid-based focal loss. But with TensorFlow, the shape of the tensor that `sigmoid_cross_entropy_with_logits` returns is different from that of `softmax_cross_entropy_with_logits`. It seems that the latter function computes a summation across the last dimension before the output, while the former does not. As a result, the additional factor of `alpha*(1-p_t)^gamma` can't be applied, since it is supposed to be done element wise on the `per_entry_cross_ent` tensor (before the summation across the last dimension).

I though of implemented a cross_entropy myself, but many TensorFlow people seem to recommend against it due to numerical instability and speed: https://github.com/tensorflow/tensorflow/issues/2462

Are there any work-arounds for this?

Thank you!

**Comment by GeorgeSeif** [comment]
As a quick update, I tried implementing the cross entropy myself as per https://github.com/tensorflow/tensorflow/issues/2462

It trained for a bit and then threw the error: `LossTensor is inf or nan. : Tensor had Inf values` which I guess was somewhat expected as the issue mentions "numerical instabilities"

**Comment by nilskk** [comment]
@GeorgeSeif Did you find a solution for implementing softmax focal loss?
